### PR TITLE
misc: Revamp exception hierarchy

### PR DESCRIPTION
**Pull request by FabioLuporini**
This was long due

Also improves handling of C-level errors

### CODE REVIEW

**Review by georgebisbas — APPROVED** [review]
Looks fine!

**Review by mloubout — APPROVED** [review]

**Review by JDBetteridge — CHANGES_REQUESTED** [review]

**Comment by JDBetteridge** [comment]
Let me know if I'm being too picky!

File: devito/operator/operator.py [hunk 1]
```diff
@@ -188,7 +189,8 @@ def _sanitize_exprs(cls, expressions, **kwargs):
 
         for i in expressions:
             if not isinstance(i, Evaluable):
-                raise InvalidOperator("`%s` is not an `Evaluable` object" % str(i))
+                raise CompilationError("`%s` is not an Evaluable object, "
+                                       "check again your Equation" % str(i))
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/operator/operator.py:193]: alright, let's start the journey with this PR

File: tests/test_buffering.py [hunk 2]
```diff
@@ -353,7 +353,7 @@ def define(self, dimensions):
 
     try:
         Operator(eqns, opt='buffering')
-    except InvalidOperator:
+    except CompilationError:
         assert True
     except:
         assert False
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply tests/test_buffering.py:359]: oh, right, that really is legacy . Will fix

[the remark this answers is not in the document]
Comment (FabioLuporini) [reply tests/test_buffering.py]: btw, in most places we're already using pytest.raises

[the remark this answers is not in the document]
Comment (JDBetteridge) [reply tests/test_buffering.py]: I thought it looked a bit odd. But in this case the failure mode wouldn't actually trigger!

**Review by JDBetteridge — APPROVED** [review]
LGTM

File: devito/operator/operator.py [hunk 3]
```diff
@@ -188,7 +189,8 @@ def _sanitize_exprs(cls, expressions, **kwargs):
 
         for i in expressions:
             if not isinstance(i, Evaluable):
-                raise InvalidOperator("`%s` is not an `Evaluable` object" % str(i))
+                raise CompilationError(f"`{i!s}` is not an Evaluable object, "
+                                       "check again your Equation")
```
Comment (EdCaunt) [inline devito/operator/operator.py:193]: Nitpick: "check your equation again" would be more natural phrasing, and the comma should probably be a semicolon

File: devito/types/dense.py [hunk 4]
```diff
@@ -1374,10 +1374,10 @@ def __init_finalize__(self, *args, **kwargs):
         available_mem = virtual_memory().available
         required_mem = np.dtype(self.dtype).itemsize * self.size
         if required_mem > available_mem:
-            warning("Trying to allocate more memory (%s) "
-                    % humanbytes(required_mem) + "for symbol %s " % self.name +
-                    "than available (%s) " % humanbytes(available_mem) +
-                    "on physical device, this will start swapping")
+            raise MemoryError(
+                f"Trying to allocate more memory ({humanbytes(required_mem)}) "
```
Comment (EdCaunt) [inline devito/types/dense.py:1378]: Honestly much better.

File: devito/operator/operator.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/operator/operator.py:193]: fixing now

**Review by mloubout — APPROVED** [review]

**Review by mloubout — APPROVED** [review]

**Pull request merged**